File: src/main/java/spring/webmvc/infrastructure/persistence/adapter/ItemRepositoryAdapter.java
package spring.webmvc.infrastructure.persistence.adapter;

import java.util.List;
import java.util.Optional;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Component;

import lombok.RequiredArgsConstructor;
import spring.webmvc.domain.model.entity.Item;
import spring.webmvc.domain.repository.ItemRepository;
import spring.webmvc.infrastructure.persistence.ItemJpaRepository;
import spring.webmvc.infrastructure.persistence.ItemQuerydslRepository;

@Component
@RequiredArgsConstructor
public class ItemRepositoryAdapter implements ItemRepository {

	private final ItemJpaRepository jpaRepository;
	private final ItemQuerydslRepository querydslRepository;

	@Override
	public Page<Item> findAll(Pageable pageable, String name) {
		return querydslRepository.findAll(pageable, name);
	}

	@Override
	public Optional<Item> findById(Long id) {
		return jpaRepository.findById(id);
	}

	@Override
	public Item save(Item item) {
		return jpaRepository.save(item);
	}

	@Override
	public List<Item> saveAll(Iterable<Item> items) {
		return jpaRepository.saveAll(items);
	}

	@Override
	public void delete(Item item) {
		jpaRepository.delete(item);
	}
}

File: spring-web-kotlin/src/main/kotlin/spring/web/kotlin/infrastructure/persistence/adapter/ItemRepositoryAdapter.kt
package spring.web.kotlin.infrastructure.persistence.adapter

import org.springframework.data.domain.Pageable
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Component
import spring.web.kotlin.domain.model.entity.Item
import spring.web.kotlin.domain.repository.ItemRepository
import spring.web.kotlin.infrastructure.persistence.ItemJpaRepository
import spring.web.kotlin.infrastructure.persistence.ItemQuerydslRepository

@Component
class ItemRepositoryAdapter(
    private val jpaRepository: ItemJpaRepository,
    private val querydslRepository: ItemQuerydslRepository,
) : ItemRepository {

    override fun findAll(pageable: Pageable, name: String?) = querydslRepository.findAll(pageable, name)

    override fun findByIdOrNull(id: Long) = jpaRepository.findByIdOrNull(id)

    override fun save(item: Item) = jpaRepository.save(item)

    override fun saveAll(items: Iterable<Item>): List<Item> = jpaRepository.saveAll(items)

    override fun delete(item: Item) {
        jpaRepository.delete(item)
    }
}